src/Experimental/traits.jl

```diff
@@ -0,0 +1,88 @@
+
+"""
+Type characterizing the mutability/immutability of a graph type.
+
+A mutable graph type `G` is expected to implement the mutability functions:
+- `add_vertex!(g::G) -> Bool`
+- `rem_vertex!(g::G, v) -> Bool`
+- `add_edge!(g::G) -> Bool`
+- `rem_edge!(g::G, i, j) -> Bool`
+- `add_vertices!(g::G, n::Integer) -> Integer` (optional)
+- `rem_vertices!(g::G, vs; keep_order::Boolean=false) -> vmap` (optional)
+"""
+abstract type GraphMutability end
+struct MutableGraph <: GraphMutability end
+struct ImmutableGraph <: GraphMutability end
+
+"""
+    IsMutable(::Type{<:AbstractGraph})
+
+Trait returning a value of `GraphMutability`, indicates whether a
+graph type is mutable or not. Mutable graph types must implement the mutability
+interface.
+"""
+IsMutable(::Type{<:AbstractGraph}) = ImmutableGraph()
+IsMutable(::Type{<:AbstractSimpleGraph}) = MutableGraph()
+IsMutable(::G) where {G <: AbstractGraph} = IsMutable(G)
+
+LightGraphs.add_vertex!(g::G) where {G <: AbstractGraph} = add_vertex!(IsMutable(G), g)
```

Comment:
Maybe we should define `add_vertex!(g::G, args...) where {G <: AbstractGraph} = add_vertex!(IsMutable(g::G), g, args...)` instead, so that graphs, which can also add vertex metadata or have other arguments there, will still work.